src/compas/datastructures/mesh/core/_mutablemapping.py

```diff
@@ -0,0 +1,149 @@
+"""
+The module contains vendored copies of a couple of classes from collections.abc
+(python standard library) with one fundamental change:
+
+ * The Mapping class does not have a __metaclass__ = ABCMeta set
+   because this causes performance issues on IronPython 2.7.x
+
+See this issue for more details: https://github.com/compas-dev/compas/issues/562
+"""
+from __future__ import absolute_import
+from __future__ import division
+from __future__ import print_function
+
+import compas
+
+__all__ = ['Mapping', 'MutableMapping']
+
+if compas.PY3:
+    import collections.abc as stdlib_collections
+else:
+    import collections as stdlib_collections
+
+
+class Mapping(object):
+    __slots__ = ()
+
+    """A Mapping is a generic container for associating key/value
+    pairs.
+    This class provides concrete generic implementations of all
+    methods except for __getitem__, __iter__, and __len__.
+    """
+
+    def get(self, key, default=None):
+        'D.get(k[,d]) -> D[k] if k in D, else d.  d defaults to None.'
+        try:
+            return self[key]
+        except KeyError:
+            return default
+
+    def __contains__(self, key):
+        try:
+            self[key]
+        except KeyError:
+            return False
+        else:
+            return True
+
+    def keys(self):
+        "D.keys() -> a set-like object providing a view on D's keys"
+        return stdlib_collections.KeysView(self)
+
+    def items(self):
+        "D.items() -> a set-like object providing a view on D's items"
+        return stdlib_collections.ItemsView(self)
+
+    def values(self):
+        "D.values() -> an object providing a view on D's values"
+        return stdlib_collections.ValuesView(self)
+
+    def __eq__(self, other):
+        if not isinstance(other, Mapping) and not isinstance(other, stdlib_collections.Mapping):
+            return NotImplemented
+        return dict(self.items()) == dict(other.items())
+
+    __reversed__ = None
+
+
+class MutableMapping(Mapping):
+
+    __slots__ = ()
+
+    """A MutableMapping is a generic container for associating
+    key/value pairs.
+    This class provides concrete generic implementations of all
+    methods except for __getitem__, __setitem__, __delitem__,
+    __iter__, and __len__.
+    """
+
+    __marker = object()
+
+    def pop(self, key, default=__marker):
+        '''D.pop(k[,d]) -> v, remove specified key and return the corresponding value.
+          If key is not found, d is returned if given, otherwise KeyError is raised.
+        '''
+        try:
+            value = self[key]
+        except KeyError:
+            if default is self.__marker:
```

Comment:
is this ever the case? won't the markers always have a different id?

Reply:
The code is fundamentally copy&pasted entirely from the python std lib, so I rather not change this behavior. That said, I think it is correct, the case can happen if there's no default (ie, default will equal to `__marker` as per `__init__`) .